rfcs/ALL/Access Control Rules.md

```diff
@@ -0,0 +1,134 @@
+        "put","get","declare_subscriber","declare_queryable"
+      ],
+      "flows":["egress","ingress"],
+      "permission": "allow",
+      "key_exprs": [
+        "test/thr"
+      ],
+      "interfaces": [
+        "lo0"
+      ]
+    },
+ ]
+}
+```
+
+The configuration has 3 fields:
+
+1. **enabled**: true/false
+2. **default_permission**: allow/deny 
+3. **rules**:  the rules set for specifying explicit allow and deny permissions
+
+The *enabled* field decides if the ACL is enabled or not. If it is set to `false`, no filtering of messages takes place and everything following that in the config is ignored.
+The *default_permission* field provides the implicit permission for the filtering, i.e., this rule applies if no other matching rule is found for an *action*. If set to `allow`, it will allow all messages to pass through unless explicitly denied in the *rules* field. If set to `deny`, it blocks all messages and only allows those that are allowed by explicit rules provided in the *rules* section. The *default_permission* always has lower priority than explicit rules provided in the *rules* section.
+
+The *rules* section itself has 5 inner fields: *actions*, *flows*, *permission*, *key_exprs*, *interfaces*
+
+* **actions**: supports four types of messsges - `put`, `get`, `declare_subscriber`, `declare_queryable`
+* **flows**: supports two values - `egress` and `ingress`
+* **permission**: supports - `allow` and `deny`
+* **key_exprs**: supports values of any key type or key-expression (set of keys) type, eg: `temp/room_1`, `temp/**` etc
+* **interfaces**: supports all possible value for network interfaces, eg: `lo`, `lo0` etc
+
+
+For each combination of *interface* + *flow* + *action*  (example:`l0`+`egress`+`put`) in the rules vector, we construct *allow* and *deny* KeTrees (key-expression tries). The *allow* KeTree is built using all the key-expressions provided in the config on which that *interface* is allowed to do that *action* on a particular *flow*. On receiving an authorization request, the key-expression in the request is matched against the appropriate KeTrees to confirm authorization.
+
+## Priority 
+For each message, the access control checks are made in the following order of priority:
+
+*explicit deny* rule > *explicit allow* rule > *default_permission* rule
+
+1. *Explicit deny* rules are given topmost priority. This means that if the key-expression in the  authorization request matches against anything in the *deny* KeTree, it will be denied immediately. If there is no match, then the *explicit allow* rules are checked.
+2. *Explicit allow* permissions come next, i.e., if request matches anything in the *allow* tree then it will be allowed.
+3. The *default_permission* value has least priority. It is applied only if there are no matches in the previous two steps.
+    
+    Note: if *default_permission* is set to `allow`, then there is no need for setting *explicit allow* rules, since they will be discarded anyways.
```

Comment:
Does the same apply for *default_permission* `deny` and explicit rules with `deny` permissions? From the pseudo-code below I understand that it does not, but we should mention it here too.
ex: "The same logic does not apply to *default_permission*=`deny` and explicit `deny`, due to permission priorities (see pseudo-code below for more detail)"